xwiki-platform-core/xwiki-platform-rest/xwiki-platform-rest-server/src/test/java/org/xwiki/rest/internal/exceptions/ExceptionExceptionMapperTest.java

```diff
@@ -0,0 +1,144 @@
+import static org.apache.commons.lang3.exception.ExceptionUtils.getStackTrace;
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertSame;
+import static org.junit.jupiter.api.Assertions.assertThrows;
+import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.verifyNoInteractions;
+import static org.mockito.Mockito.when;
+
+/**
+ * Test of {@link ExceptionExceptionMapper}.
+ *
+ * @version $Id$
+ * @since 13.8RC1
+ * @since 13.7.1
+ * @since 13.4.4
+ */
+@ComponentTest
+class ExceptionExceptionMapperTest
+{
+    @RegisterExtension
+    LogCaptureExtension logCapture = new LogCaptureExtension(LogLevel.ERROR);
+
+    @InjectMockComponents
+    private ExceptionExceptionMapper exceptionExceptionMapper;
+
+    @MockComponent
+    private TemplateManager templateManager;
+
+    @MockComponent
+    private ScriptContextManager scriptContextManager;
+
+    @MockComponent
+    private ContextualLocalizationManager contextLocalization;
+
+    @Mock
+    private ScriptContext scriptContext;
+
+    @Test
+    void toResponseUnchecked() throws Exception
+    {
+        IOException cause = new IOException("file not found");
+        String expectedMessage = String.format(
+            "No ExceptionMapper was found for [java.io.IOException: file not found].\n%s",
+            getStackTrace(cause));
+
+        when(this.contextLocalization.getTranslationPlain("rest.exception.noMapper", cause))
+            .thenReturn("No ExceptionMapper was found for [java.io.IOException: file not found].");
+        when(this.scriptContextManager.getScriptContext()).thenReturn(this.scriptContext);
+        when(this.templateManager.render("rest/exception.vm")).thenReturn(expectedMessage);
+
+        Response response = this.exceptionExceptionMapper.toResponse(cause);
+
+        assertEquals(expectedMessage, response.getEntity());
+        assertEquals(500, response.getStatus());
+        assertEquals(TEXT_HTML_TYPE, response.getMetadata().get(CONTENT_TYPE).get(0));
+        assertEquals(1, this.logCapture.size());
+        assertEquals("A REST endpoint failed with an unmapped exception.", this.logCapture.getMessage(0));
+        assertEquals("java.io.IOException", this.logCapture.getLogEvent(0).getThrowableProxy().getClassName());
+        assertEquals("file not found", this.logCapture.getLogEvent(0).getThrowableProxy().getMessage());
+        assertEquals(ERROR, this.logCapture.getLogEvent(0).getLevel());
+
+        verify(this.scriptContext).setAttribute("cause", cause, ScriptContext.ENGINE_SCOPE);
+        verifyNoInteractions(this.contextLocalization);
+    }
+
+    @Test
+    void toResponseUncheckedTemplateRenderError() throws Exception
+    {
+        IOException cause = new IOException("file not found");
+        String expectedMessage = String.format(
+            "No ExceptionMapper was found for [java.io.IOException: file not found].\n%s",
+            getStackTrace(cause));
```

Comment:
Personally I don't like tests that are too clever. Small code duplication is better in tests in general for me when it makes the test more readable and easier to understand. Here I find it hard to see the full error message we're expecting.

Reply:
I get your point. I wrote it that way because I'm did not find an simpler way to assert that the log contains a well-formed stacktrace while keeping the test portable (for instance I'm not sure the stacktrace will stay the same on the CI, or if we update one ouf our test libraries).